src/main/java/hudson/remoting/AtmostOneThreadExecutor.java

```diff
@@ -71,10 +71,11 @@ public boolean isTerminated() {
 
     public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
         synchronized (q) {
-            long start = System.currentTimeMillis();
-            long end = start+unit.toMillis(timeout);
-            while (isAlive() && System.currentTimeMillis()<end) {
-                q.wait(end-System.currentTimeMillis());
+            long now = System.nanoTime();
+            long end = now + unit.toNanos(timeout);
+            while (isAlive() && (end - now > 0L)) {
```

Comment:
Note that for `timeout == 0`, this will now `wait(0)`, whereas the original code would have skipped the loop altogether. Probably does not matter, just FYI.